<a href="https://colab.research.google.com/github/Andre-Gomez-Serrano/2.2-Actividad-2-An-lisis-de-datos-con-pandas/blob/main/Actividad3BDManipulacion_A01733624.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: Carlos André Gómez Serrano
*   MATRÍCULA: A01733624


---

En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

* `movie`: Información general de las películas.

* `tv_show`: Información general de los programas de televisión.

* `season`: Datos de las temporadas asociadas a cada programa de TV.

* `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (*driver*) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.

In [ ]:
pip install pymysql

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.7/46.7 kB 1.8 MB/s eta 0:00:00


`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar *commits* y *rollbacks* de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”

In [9]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.

In [10]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine('mysql+pymysql://mnaTC5053User:mnaTC5053Pass!@20.127.169.219:3306/Netflix', pool_recycle=3600)
conn = db.connect()

Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```

1.	Extrae toda la información de las películas que duran más de 5 horas.

In [19]:
query = sqla.text("""
  SELECT *
  FROM movie
  WHERE runtime > 300
""")
pd.read_sql(query, conn)

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,5793,2024-01-01,2024-01-01,b'\x00',None,日本統一シリーズ: 映画シリーズ,None,3892,Nihontouitsu Series: Film Series
1,5794,2024-01-01,2024-01-01,b'\x00',None,釣りバカ日誌: 映画シリーズ,None,2120,Free and Easy Series: Film Series
2,5874,2024-01-01,2024-01-01,b'\x01',None,None,2021-08-06,312,Navarasa: Limited Series
3,9729,2024-01-01,2024-01-01,b'\x00',None,織田同志会 織田征仁: 映画シリーズ,None,710,Seiji Oda: Film Series
4,9730,2024-01-01,2024-01-01,b'\x00',None,キングダム～首領になった男～: 映画シリーズ,None,427,Kingdom ~ The Man Who Became the Top ~: Film S...


2.	¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.

In [23]:
query = sqla.text("""
  SELECT
    100.0 * SUM(CASE WHEN NOT available_globally THEN 1 ELSE 0 END)
          / COUNT(available_globally) AS porcentaje_solo_eu
  FROM movie
  WHERE available_globally IS NOT NULL
""")
pd.read_sql(query, conn)

,porcentaje_solo_eu
0,83.57175


3.	¿Cuáles son los idiomas o regiones originales en la tabla de películas?
* ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)

In [27]:
#Idiomas o regiones originales en la tabla de películas
query = sqla.text("""
  SELECT DISTINCT locale
  FROM movie
  ORDER BY locale
""")
pd.read_sql(query, conn)

,locale
0,None
1,en


In [28]:
#Cuántos registros tiene el campo "locale" con valor NULL.
query = sqla.text("""
  SELECT COUNT(*) AS registros_null
  FROM movie
  WHERE locale IS NULL
""")
pd.read_sql(query, conn)

,registros_null
0,11255


4.	Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.
* Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
*  ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
* Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.

In [30]:
#Películas que tienen títulos diferentes a "title" y "original_title"
query = sqla.text("""
  SELECT COUNT(*) AS titulos_diferentes
  FROM movie
  WHERE NOT (title <=> original_title)
""")
pd.read_sql(query, conn)

,titulos_diferentes
0,11770


In [31]:
#¿Coinciden los resultados (cantidad de NULL y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
query = sqla.text("""
  SELECT COUNT(*) AS titulos_diferentes
  FROM movie
  WHERE NOT (title <=> original_title) AND locale IS NULL
""")
pd.read_sql(query, conn)

,titulos_diferentes
0,11238


In [33]:
query = sqla.text("""
  SELECT
    (locale IS NULL)                 AS locale_es_null,
    (NOT (title <=> original_title)) AS titulos_diferentes,
    COUNT(*)                         AS total
  FROM movie
  GROUP BY 1, 2
  ORDER BY 1, 2
""")
pd.read_sql(query, conn)

,locale_es_null,titulos_diferentes,total
0,0,1,532
1,1,0,17
2,1,1,11238


¿Coinciden los resultados (cantidad de NULL y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
Hay 11,255 registros con locale NULL y 11,770 películas con title distinto de original_title, una diferencia de 515. Los registros restantes son de dos tipos. Primero, 532 películas con locale informado (no NULL) que también tienen títulos diferentes; no existe ninguna película con locale informado y títulos iguales. Segundo, 17 películas con locale NULL cuyos títulos MySQL considera iguales. Estas 17 se caracterizan por tener títulos que solo difieren en acentos o diacríticos (por ejemplo, "Recep Ivedik 7" frente a "Recep İvedik 7", o "Totem" frente a "Tótem"), casi todas sin fecha de estreno (solo una la tiene) y la mayoría con available_globally en falso (15 de 17).

Finalmente, concluye si la suposición de que los valores NULL en locale indican que la película está en otro idioma es válida.

La suposición es válida en teoría. De las 11,255 películas con locale NULL, 11,238 tienen un título diferente al original. Las 17 restantes son una excepción aparente: sus títulos solo difieren en acentos, que la comparación de texto de MySQL ignora, y corresponden a títulos de origen turco, español, francés y rumano, entre otros. Por lo tanto, un locale NULL indica de forma consistente una película cuyo idioma original no es el inglés.

5.	Determina el título de la película que ha permanecido más tiempo en el top 10.

In [39]:
query = sqla.text("""
  SELECT m.title,
         MAX(v.cumulative_weeks_in_top10) AS semanas_en_top10
  FROM view_summary v
  INNER JOIN movie m ON m.id = v.movie_id
  GROUP BY m.id, m.title
  ORDER BY semanas_en_top10 DESC
  LIMIT 1
""")
pd.read_sql(query, conn)

,title,semanas_en_top10
0,The Boss Baby,22


6.	Identifica los 5 programas de TV con mayor cantidad de temporadas.

In [40]:
conn.rollback()
query = sqla.text("""
  SELECT t.title,
         COUNT(s.id) AS total_temporadas
  FROM tv_show t
  INNER JOIN season s ON s.tv_show_id = t.id
  GROUP BY t.id, t.title
  ORDER BY total_temporadas DESC
  LIMIT 5
""")
pd.read_sql(query, conn)

,title,total_temporadas
0,Grey's Anatomy,20
1,Naruto Shippuden,20
2,Heartland (2007),17
3,It's Always Sunny in Philadelphia,16
4,NCIS,15


7.	¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?

In [41]:
conn.rollback()
query = sqla.text("""
  SELECT DISTINCT start_date, end_date
  FROM view_summary
  WHERE duration = 'SEMI_ANNUALLY'
  ORDER BY start_date
""")
pd.read_sql(query, conn)

,start_date,end_date
0,2023-07-01,2023-12-31
1,2024-01-01,2024-06-30


8.	Ordena las temporadas de *Grey's Anatomy* según la cantidad de vistas registradas en el primer período semestral de 2024.
* ¿Cómo interpretarías los resultados obtenidos?

In [42]:
#Ordena las temporadas de Grey's Anatomy
conn.rollback()
query = sqla.text("""
  SELECT t.title AS programa,
         s.season_number,
         s.title AS temporada,
         v.views
  FROM view_summary v
  INNER JOIN season s ON s.id = v.season_id
  INNER JOIN tv_show t ON t.id = s.tv_show_id
  WHERE t.title LIKE 'Grey%Anatomy%'
    AND v.duration = 'SEMI_ANNUALLY'
    AND v.start_date = '2024-01-01'
  ORDER BY v.views DESC
""")
pd.read_sql(query, conn)

,programa,season_number,temporada,views
0,Grey's Anatomy,1,Grey's Anatomy: Season 1,3600000
1,Grey's Anatomy,2,Grey's Anatomy: Season 2,3100000
2,Grey's Anatomy,3,Grey's Anatomy: Season 3,2900000
3,Grey's Anatomy,5,Grey's Anatomy: Season 5,2900000
4,Grey's Anatomy,4,Grey's Anatomy: Season 4,2900000
5,Grey's Anatomy,6,Grey's Anatomy: Season 6,2800000
6,Grey's Anatomy,7,Grey's Anatomy: Season 7,2700000
7,Grey's Anatomy,8,Grey's Anatomy: Season 8,2600000
8,Grey's Anatomy,9,Grey's Anatomy: Season 9,2500000
9,Grey's Anatomy,10,Grey's Anatomy: Season 10,2400000


9.	Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.

In [44]:
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

In [46]:
import pandas as pd
import sqlalchemy as sqla
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)
df = (
    view_summary
    .merge(season, left_on='season_id', right_on='id',
           how='inner', suffixes=('_vista', '_temporada'))
    .merge(tv_show, left_on='tv_show_id', right_on='id',
           how='inner', suffixes=('_temporada', '_programa'))
)
resultado = df[
    df['title_programa'].str.contains("Grey's Anatomy", case=False, na=False)
    & (df['duration'] == 'SEMI_ANNUALLY')
    & (pd.to_datetime(df['start_date']) == '2024-01-01')
]
resultado = (
    resultado[['title_programa', 'season_number', 'title_temporada', 'views']]
    .sort_values('views', ascending=False)
    .reset_index(drop=True)
)

resultado

,title_programa,season_number,title_temporada,views
0,Grey's Anatomy,1.0,Grey's Anatomy: Season 1,3600000
1,Grey's Anatomy,2.0,Grey's Anatomy: Season 2,3100000
2,Grey's Anatomy,3.0,Grey's Anatomy: Season 3,2900000
3,Grey's Anatomy,5.0,Grey's Anatomy: Season 5,2900000
4,Grey's Anatomy,4.0,Grey's Anatomy: Season 4,2900000
5,Grey's Anatomy,6.0,Grey's Anatomy: Season 6,2800000
6,Grey's Anatomy,7.0,Grey's Anatomy: Season 7,2700000
7,Grey's Anatomy,8.0,Grey's Anatomy: Season 8,2600000
8,Grey's Anatomy,9.0,Grey's Anatomy: Season 9,2500000
9,Grey's Anatomy,10.0,Grey's Anatomy: Season 10,2400000


`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.

En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [1]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.

In [4]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [15]:
# Debes descargar el archivo de credenciales "consultancy.json" (disponible en las instrucciones de Canvas) y ubicarte en la carpeta donde lo almacenaste
import os
DIR = "/content/drive/MyDrive/Colab Notebooks/MNA/TC5053 - Ciencia y analítica de datos/Actividad3_BD_Manipulacion"
os.chdir("/content/drive/MyDrive/Colab Notebooks")


In [16]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

10.	Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado

In [18]:
import pandas as pd
docs = db.collection('EMPLOYEE').stream()
data = []
for doc in docs:
    registro = doc.to_dict()
    registro['id'] = doc.id
    data.append(registro)
employee = pd.DataFrame(data).set_index('id')

employee

,emp_fname,emp_lname,emp_hiredate
id,,,
8LcLuxVHGAd2d9IQc5jF,David,Senior,1989-07-12 06:00:00+00:00
Fzd60D6Z2CU4n0wVV8YN,William,Smithfield,2004-06-04 05:00:00+00:00
lX5xuQ5w3i6ib2ExccWY,John,News,2000-11-08 06:00:00+00:00
yocFj2lichOkbAj9NBfp,June,Arbough,1996-12-01 06:00:00+00:00


In [ ]:
firebase_admin.delete_app(firebase_admin.get_app())

---

**Declaración de uso de IA**


*   Claude.IA. (2026). *Claude.IA (basado en Sonnet 5.5)* [Modelo de lenguaje grande], utilizado para generación de código y depuración. https://claude.ai/new

---